File: src/MagnusAddon/jaslib/ui/web/sentence/compound_part_viewmodel.py
from __future__ import annotations

from typing import TYPE_CHECKING

from ankiutils import app
from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]
from configuration.settings import Settings
from typed_linq_collections.collections.q_set import QSet

if TYPE_CHECKING:
    from jaslib.note.jpnote import NoteId
    from jaslib.note.sentences.sentence_configuration import SentenceConfiguration
    from jaslib.note.vocabulary.vocabnote import VocabNote
    from jaslib.ui.web.sentence.match_viewmodel import MatchViewModel

class CompoundPartViewModel(Slots):
    def __init__(self, vocab_note: VocabNote, depth: int, config: SentenceConfiguration) -> None:
        self.vocab_note: VocabNote = vocab_note
        self.depth: int = depth
        self.question: str = vocab_note.get_question()
        self.answer: str = vocab_note.get_answer()
        self.readings: str = ", ".join(vocab_note.readings.get())
        self.audio_path: str = vocab_note.audio.get_primary_audio_path()
        self.meta_tags_html: str = vocab_note.meta_data.meta_tags_html(no_sentense_statistics=True)
        self.display_readings: bool = self.question != self.readings
        self.is_highlighted: bool = self.question in config.highlighted_words

        self.meta_tags_string: str = " ".join(vocab_note.get_meta_tags())
        self.meta_tags_string += f""" depth_{depth}"""
        self.meta_tags_string += " highlighted" if self.is_highlighted else ""

    @classmethod
    def get_compound_parts_recursive(cls, match_viewmodel: MatchViewModel, vocab_note: VocabNote, config: SentenceConfiguration, depth: int = 0, visited: QSet[NoteId] | None = None) -> list[CompoundPartViewModel]:
        if not Settings.hide_all_compounds():
            if not Settings.show_compound_parts_in_sentence_breakdown(): return []
            if visited is None: visited = QSet()
            if vocab_note.get_id() in visited: return []

            visited.add(vocab_note.get_id())

            result: list[CompoundPartViewModel] = []

            for part in vocab_note.compound_parts.primary_parts_notes():  # ex_sequence.flatten([app.col().vocab.with_form_prefer_exact_match(part) for part in vocab_note.compound_parts.primary()])
                wrapper = CompoundPartViewModel(part, depth, config)
                result.append(wrapper)
                nested_parts = cls.get_compound_parts_recursive(match_viewmodel, part, config, depth + 1, visited)
                result.extend(nested_parts)

            return result
        match = match_viewmodel.match
        if match.inspector.is_ichidan_covering_godan_potential:
            godan_base = match.word.start_location.token.base_form
            godan_potential_part_base = match.word.end_location.token.base_form
            godan = app.col().vocab.with_form_prefer_disambiguation_name_or_exact_match(godan_base)
            godan_potential = app.col().vocab.with_form_prefer_disambiguation_name_or_exact_match(godan_potential_part_base)
            if godan and godan_potential:
                return [(CompoundPartViewModel(godan[0], depth, config)),
                        (CompoundPartViewModel(godan_potential[0], depth, config))]
            return []
        # we may still have parts if janome tokenizes a word we consider a compound as a single token
        return (vocab_note.compound_parts.primary_parts_notes()
                .select(lambda _part: CompoundPartViewModel(_part, depth, config)).to_list())

File: src/src_dotnet/JAStudio.Core/UI/Web/Sentence/CompoundPartViewModel.cs
using System.Collections.Generic;
using System.Linq;
using JAStudio.Core.Note;
using JAStudio.Core.Note.Sentences;
using JAStudio.Core.Note.Vocabulary;
namespace JAStudio.Core.UI.Web.Sentence;

public class CompoundPartViewModel
{
    public VocabNote VocabNote { get; }
    public int Depth { get; }
    public string Question { get; }
    public string Answer { get; }
    public string Readings { get; }
    public string AudioPath { get; }
    public string MetaTagsHtml { get; }
    public bool DisplayReadings { get; }
    public bool IsHighlighted { get; }
    public string MetaTagsString { get; }

    public CompoundPartViewModel(VocabNote vocabNote, int depth, SentenceConfiguration config)
    {
        VocabNote = vocabNote;
        Depth = depth;
        Question = vocabNote.GetQuestion();
        Answer = vocabNote.GetAnswer();
        Readings = string.Join(", ", vocabNote.GetReadings());
        AudioPath = vocabNote.Audio.GetPrimaryAudioPath();
        MetaTagsHtml = vocabNote.MetaData.MetaTagsHtml(noSentenceStatistics: true);
        DisplayReadings = Question != Readings;
        IsHighlighted = config.HighlightedWords.Contains(Question);

        MetaTagsString = string.Join(" ", vocabNote.GetMetaTags());
        MetaTagsString += $" depth_{depth}";
        MetaTagsString += IsHighlighted ? " highlighted" : "";
    }

    public static List<CompoundPartViewModel> GetCompoundPartsRecursive(
        MatchViewModel matchViewModel,
        VocabNote vocabNote,
        SentenceConfiguration config,
        int depth = 0,
        HashSet<long>? visited = null)
    {
        if (!TemporaryServiceCollection.Instance.Settings.HideAllCompounds())
        {
            if (!TemporaryServiceCollection.Instance.Settings.ShowCompoundPartsInSentenceBreakdown()) return [];
            visited ??= [];
            if (visited.Contains(vocabNote.GetId())) return [];

            visited.Add(vocabNote.GetId());

            var result = new List<CompoundPartViewModel>();

            foreach (var part in vocabNote.CompoundParts.PrimaryPartsNotes())
            {
                var wrapper = new CompoundPartViewModel(part, depth, config);
                result.Add(wrapper);
                var nestedParts = GetCompoundPartsRecursive(matchViewModel, part, config, depth + 1, visited);
                result.AddRange(nestedParts);
            }

            return result;
        }

        var match = matchViewModel.Match;
        if (match.Inspector.IsIchidanCoveringGodanPotential)
        {
            var godanBase = match.Word.StartLocation.Token.BaseForm;
            var godanPotentialPartBase = match.Word.EndLocation.Token.BaseForm;
            var godan = TemporaryServiceCollection.Instance.App.Col().Vocab.WithFormPreferDisambiguationNameOrExactMatch(godanBase);
            var godanPotential = TemporaryServiceCollection.Instance.App.Col().Vocab.WithFormPreferDisambiguationNameOrExactMatch(godanPotentialPartBase);
            if (godan.Any() && godanPotential.Any())
            {
                return
                [
                   new CompoundPartViewModel(godan.First(), depth, config),
                   new CompoundPartViewModel(godanPotential.First(), depth, config)
                ];
            }
            return [];
        }

        // We may still have parts if janome tokenizes a word we consider a compound as a single token
        return vocabNote.CompoundParts.PrimaryPartsNotes()
            .Select(part => new CompoundPartViewModel(part, depth, config))
            .ToList();
    }
}
